slang-checks/src/main/java/com/sonarsource/slang/checks/TooDeeplyNestedStatementsCheck.java

```diff
@@ -0,0 +1,123 @@
+import java.util.List;
+
+@Rule(key = "S134")
+public class TooDeeplyNestedStatementsCheck implements SlangCheck {
+  private static final int DEFAULT_MAX_DEPTH = 3;
+
+  @RuleProperty(
+      key = "max",
+      description = "Maximum allowed control flow statement nesting depth",
+      defaultValue = "" + DEFAULT_MAX_DEPTH)
+
+  public int max;
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(IfTree.class, this::checkNestedDepth);
+    init.register(LoopTree.class, this::checkNestedDepth);
+    init.register(MatchTree.class, this::checkNestedDepth);
+    init.register(ExceptionHandlingTree.class, this::checkNestedDepth);
+  }
+
+  private void checkNestedDepth(CheckContext ctx, Tree tree) {
+    if (isElseIfStatement(ctx.parent(), tree)) {
+      return;
+    }
+
+    Iterator<Tree> iterator = ctx.ancestors().iterator();
+    Deque<Token> nestedParentNodes = new LinkedList<>();
+    Tree last = tree;
+
+    while (iterator.hasNext()) {
+      Tree parent = iterator.next();
+      if (isElseIfStatement(parent, last) && !nestedParentNodes.isEmpty()) {
+        nestedParentNodes.removeLast();
+      }
+
+      if (parent instanceof LoopTree || parent instanceof ExceptionHandlingTree || parent instanceof IfTree || parent instanceof MatchTree) {
+        nestedParentNodes.addLast(getNodeToHighlight(parent));
+      }
+
+      if (nestedParentNodes.size() > max) {
+        return;
+      }
+      last = parent;
+    }
+
+    if (nestedParentNodes.size() == max) {
+      reportIssue(ctx, tree, nestedParentNodes);
+    }
+  }
+
+  private static boolean isElseIfStatement(@Nullable Tree parent, @Nullable Tree tree) {
+    return tree instanceof IfTree && parent instanceof IfTree && tree.equals(((IfTree) parent).elseBranch());
+  }
+  
+  private void reportIssue(CheckContext ctx, Tree statement, Deque<Token> nestedStatements) {
+    String message = String.format("Refactor this code to not nest more than %s control flow statements.", max);
+    List<SecondaryLocation> secondaryLocations = new ArrayList<>(nestedStatements.size());
+    int nestedDepth = 0;
+
+    while (!nestedStatements.isEmpty()) {
+      nestedDepth++;
+      String secondaryLocationMessage = String.format("Nesting depth %s", nestedDepth);
+      secondaryLocations.add(new SecondaryLocation(nestedStatements.removeLast().textRange(), secondaryLocationMessage));
+    }
+
+    Token nodeToHighlight = getNodeToHighlight(statement);
+    ctx.reportIssue(nodeToHighlight, message, secondaryLocations);
+  }
+
+  private static Token getNodeToHighlight(Tree tree) {
+    if (tree instanceof IfTree) {
+      return ((IfTree) tree).ifKeyword();
+    } else if (tree instanceof MatchTree) {
+      return ((MatchTree) tree).keyword();
+    } else if (tree instanceof ExceptionHandlingTree) {
+      return ((ExceptionHandlingTree) tree).tryKeyword();
+    } else if (tree instanceof LoopTree) {
+      return ((LoopTree) tree).keyword();
+    } else {
```

Comment:
I don't think it can ever happen to reach this clause given the logic in the code